src/util/expr_iterator.h

```diff
@@ -0,0 +1,247 @@
+/*******************************************************************\
+
+ Module: exprt iterator module
+
+ Author: DiffBlue Limited. All rights reserved.
+
+\*******************************************************************/
+
+#ifndef CPROVER_UTIL_EXPR_ITERATOR_H
+#define CPROVER_UTIL_EXPR_ITERATOR_H
+
+#include <iterator>
+#include <functional>
+#include <set>
+#include <algorithm>
+#include "expr.h"
+#include "invariant.h"
+
+/// \file Forward depth-first search iterators
+/// These iterators' copy operations are expensive, so use auto&, and avoid
+/// std::next(), std::prev() etc.
+///
+/// Incrementable only by calling ++it (to avoid accidental copying, as
+/// this iterator is not cheap)
+///
+/// Non-const iterators dereference to const exprt (for use with STL
+/// algorithms) but have an extra .mutate() method. That method is used
+/// to access non-const exprt reference but is an expensive operation
+
+/// Naive depth-first-search iterator with .mutate method for
+/// modifying exprt under the iterator.
+class depth_iteratort;
```

Comment:
Depends on whether it's forbidden to have forward declarations for easier reading.

It shows what is in the file without having to go through the entire source code. If that's bad, I'll move them around and will never attempt it again (originally I wanted to expand on that idea and move iterator definitions out of iterator).